From 0aef272076ad875a3a1f6b7f9e330a31949700bc Mon Sep 17 00:00:00 2001
From: Yusuke Suzuki <ysuzuki@apple.com>
Date: Mon, 7 Apr 2025 16:16:34 -0700
Subject: [PATCH] [JSC] GetByIdModeMetadata's mode needs to be set when not
 using 64bit pointer https://bugs.webkit.org/show_bug.cgi?id=289017
 rdar://145852436

Reviewed by Mark Lam.

On environment having 32bit pointer or non-little-endian, mode needs to
be set.

* Source/JavaScriptCore/bytecode/GetByIdMetadata.h:
(JSC::GetByIdModeMetadata::setProtoLoadMode):

Originally-landed-as: 289651.211@safari-7621-branch (38a19eb3b7c0). rdar://148052940
Canonical link: https://commits.webkit.org/293387@main
---
 Source/JavaScriptCore/bytecode/GetByIdMetadata.h | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/Source/JavaScriptCore/bytecode/GetByIdMetadata.h b/Source/JavaScriptCore/bytecode/GetByIdMetadata.h
index 0f1489cb0861..d504c4a8aa3e 100644
--- a/Source/JavaScriptCore/bytecode/GetByIdMetadata.h
+++ b/Source/JavaScriptCore/bytecode/GetByIdMetadata.h
@@ -151,8 +151,12 @@ inline void GetByIdModeMetadata::setArrayLengthMode()
 
 inline void GetByIdModeMetadata::setProtoLoadMode(Structure* structure, PropertyOffset offset, JSObject* cachedSlot)
 {
+#if CPU(LITTLE_ENDIAN) && CPU(ADDRESS64)
     // We rely on ProtoLoad being 0, or else the high bits of the pointer would write the wrong mode and hit count
     static_assert(!static_cast<std::underlying_type_t<GetByIdMode>>(GetByIdMode::ProtoLoad)); // In 64bit architecture, this field is shared with protoLoadMode.cachedSlot.
+#else
+    mode = GetByIdMode::ProtoLoad;
+#endif
 
     protoLoadMode.structureID = structure->id();
     protoLoadMode.cachedOffset = offset;
-- 
2.34.1

